### PR TITLE
Hotfix/2729 fetch translated exercises part 2

### LINKED ISSUE #2729: Fetch translated exercises
Right now we only support officially internationalized KA sites like fr, pt-br and es. We wanna expand that to any user-translated exercises like de. We should also fetch translations for the Perseus exercises.

### PR DESCRIPTION
**Pull request by cpauya**
Hi @aronasorman - this fixes #2729 with these notes from @jesumer 

**Central Server Requirements**

1. [Include the .po files from Khan Academy's Crowdin account during "update_language_packs".](https://github.com/fle-internal/ka-lite-central/pull/226)
1. This is required to get the latest translations but the above should do for our test
    * [Issue fix by adding --rebuild make option into update_language_packs command](https://github.com/fle-internal/ka-lite-central/pull/230)
1. make sure it's the same version as distributed

**Distributed Server Requirements**

1. make sure it's the same version as central

**To test**

At Central server:

1. Run `python manage.py update_language_packs -l de --no-srts --no-dubbed --no-kalite-trans --rebuild` command.
1. Check if this will get the proper updated mo.files from CrowdIn.(we can check this out to centralserver/_crowdin_cache zip file).

At Distributed:

1. If it is successful then go to the distributed browser.(NOTE: Make sure you've done the "setup" already).
1. Run `python manage.py languagepackdownload -l de --commandline`.
1. Go to the language page and set the language code to `de`.
1. Go to the exercises and try to test on some of the exercises.
1. Check the translated words.

**To test thru the Python shell** 
This is taken from https://github.com/fle-internal/ka-lite-central/pull/226#issuecomment-75382550

```python
from django.utils.translation import activate, ugettext as _

activate("de")
s = "What is <code>X</code>?"
_(s)
```

Which should display this:

```python
>>> from django.utils.translation import activate, ugettext as _
>>>
>>> activate("de")
>>> s = "What is <code>X</code>?"
>>> _(s)
'Was ist <code>X</code>?'
>>>
```

References:

1. [Official Khan Academy German translations on CrowdIn](https://crowdin.com/project/khanacademy/de)
2. [learn.math.algebra-basics.exercises.pot - 43%](https://crowdin.com/translate/khanacademy/30006/enus-de) -- Search for `What is <code>X</code>?` text as per example above.

### CODE REVIEW

File: kalite/topic_tools/__init__.py [hunk 1]
```diff
@@ -491,10 +491,35 @@ def get_assessment_item_data(request, assessment_item_id=None):
     if not assessment_item:
         return None
 
-    # TODO (rtibbles): Enable internationalization for the assessment_items.
+    # Enable internationalization for the assessment_items.
+    try:
+        item_data = json.loads(assessment_item['item_data'])
+        question_content = _(item_data['question']['content'])
+        answerarea_content = item_data['answerArea']['options']['content']
+        if answerarea_content != "":
```
Comment (MCGallaspy) [inline kalite/topic_tools/__init__.py]: The empty string is falsy, so you could equivalently write: `if not answerarea_content:`

Comment (MCGallaspy) [inline kalite/topic_tools/__init__.py]: Also, why do you expect that this item may be empty, but the others (like `item_data['question']['content']`) are not?

File: kalite/topic_tools/__init__.py [hunk 2]
```diff
@@ -491,10 +491,35 @@ def get_assessment_item_data(request, assessment_item_id=None):
...
+        # Loop over contents for multiple hints.
+        # Pickup the wrapped strings and append to hints dict
+        new_hints = []
+        for contents_dict in item_data['hints']:
+            content_value = contents_dict.get('content')
+            converted_content = _(content_value)
+            contents_dict['content'] = converted_content
+            new_hints.append(contents_dict)
+
+        item_data['hints'] = new_hints
```
Comment (MCGallaspy) [inline kalite/topic_tools/__init__.py]: I think you could replace 504 to 511 with this:
```python
item_data['hints'] = map(lambda h: h['content'] = _(h.get('content')), item_data['hints'])
```

**Comment by aronasorman** [comment]
I took quite some time trying to find a perseus exercise that's at least partially translated in KA.

Exercise: Linear Algebra -> Unit Vectors 

screen shot 2015-03-10 at 11 24 15 am


Notice the instructions to the right (Fulle die Lucken)

Which is present in the assessment item data:
screen shot 2015-03-10 at 11 27 59 am

So I can confirm that this works. Although @cpauya maybe you have a better example for everyone to see.

File: kalite/topic_tools/__init__.py [hunk 3]
```diff
@@ -491,10 +491,35 @@ def get_assessment_item_data(request, assessment_item_id=None):
...
+        item_data['question']['content'] = question_content
+        item_data['answerArea']['options']['content'] = answerarea_content
+        # dump the data for a proper json format.
+        assessment_item['item_data'] = json.dumps(item_data)
+
+    except KeyError:
+        logging.error("There is something wrong with the format of the assessment items:%s" % KeyError)
```
Comment (MCGallaspy) [inline kalite/topic_tools/__init__.py]: "something wrong" is a bit vague, can you change it to "Assessment item did not have expected key"? Also, this probably doesn't have the output you expect (`KeyError` will refer to the class `KeyError` and not the instance that you're catching). Plus `KeyError`s are not very verbose.
```python
except KeyError as e:
    logging.error("Assessment item did not have the expected key %s. Assessment item: \n %s" % (e, assessment_item))
```

File: kalite/topic_tools/__init__.py (hunk 2 above)
Comment (aronasorman) [inline kalite/topic_tools/__init__.py]: @MCGallaspy that's not technically correct, as assignment is a statement and won't return any value, which is what a lambda expression expects. I'm not sure if there's a succinct way to just change one element of a dictionary.

**Comment by aronasorman** [comment]
Addressed some of mike's issues. Just waiting for tests to pass, and then I'm merging this.

**Comment by cpauya** [comment]
Hi @aronasorman - thanks for the fixes.

Will try to find an exercise and post here with screenshots.

We will need this issue merged to central though to get the updated translations: [Issue fix by adding --rebuild make option into update_language_packs command](https://github.com/fle-internal/ka-lite-central/pull/230)

**Comment by cpauya** [comment]
@aronasorman Sadly I can't try the exercise you have above because it's greyed-out on my sidebar topics and loading the entry loads a "Content was not found..." message.

ka-lite

I'm waiting for `bin/kalite manage unpack_assessment_zip http://eslgenie.com/media/assessment_item_resources.zip -f` to finish to see if an updated `assessmentitems.json` will fix it.

**Comment by cpauya** [comment]
@aronasorman - I was able to reproduce your string above (see screenshot) so executing `bin/kalite manage unpack_assessment_zip http://eslgenie.com/media/assessment_item_resources.zip -f` helped!  Thanks for the tip.

screenshot 2015-03-11 23 13 16

Found the discrepancy between the `asssessmentitems.json` strings and the translation strings from CrowdIn at `django.mo`.  I used this exercise url: http://127.0.0.1:8008/learn/khan/math/early-math/cc-early-math-counting-topic/cc-early-math-counting/count-to-100/

**Sample script**
```python
from django.utils.translation import activate, ugettext as _

activate("de")

# 1. Sample string to check that we are indeed translating to German.
s = "What is <code>X</code>?"
_(s)

# 2. String from `assessmentitems.json` API call to http://127.0.0.1:8008/api/assessment_item/x83e816580f82fd5f/
s = "Let's count by tens.\n\n![](/content/khan/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png)\n\n**What number did we miss?**\n\n \n\n[[\u2603 radio 1]]"
_(s)

# 3. This is the string from the CrowdIn file learn.math.early-math.exercises-de.po:31062
s = "Let's count by tens.\n\n![](https://ka-perseus-graphie.s3.amazonaws.com/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png)\n\n**What number did we miss?**\n\n \n\n[[☃ radio 1]]"
_(s)
```

And here's the result

```python
>>> from django.utils.translation import activate, ugettext as _
>>>
>>> activate("de")
>>>
>>> # 1. Sample string to check that we are indeed translating to German.
>>> s = "What is <code>X</code>?"
>>> _(s)
'Was ist <code>X</code>?'
>>>
>>> # 2. String from `assessmentitems.json` API call to http://127.0.0.1:8008/api/assessment_item/x83e816580f82fd5f/
>>> s = "Let's count by tens.\n\n![](/content/khan/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png)\n\n**What number did we miss?**\n\n \n\n[[\u2603 radio 1]]"
>>> _(s)
u"Let's count by tens.\n\n![](/content/khan/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png)\n\n**What number did we miss?**\n\n \n\n[[\\u2603 radio 1]]"
>>>
>>> # 3. This is the string from the CrowdIn file learn.math.early-math.exercises-de.po:31062
>>> s = "Let's count by tens.\n\n![](https://ka-perseus-graphie.s3.amazonaws.com/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png)\n\n**What number did we miss?**\n\n \n\n[[☃ radio 1]]"
>>> _(s)
'Lass uns in Zehnerschritten z\xc3\xa4hlen. \n\n ![](https://ka-perseus-graphie.s3.amazonaws.com/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png) \n\n **Welche Zahl fehlt?** \n\n\n\n [[\xe2\x98\x83 radio 1]]'
```

**Question**

1. Based on comment 2 above - The string from `asssessmentitems.json` differs from the string that's on the `django.mo` translation file:

    1.1.The `/content/khan/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png` versus the `https://ka-perseus-graphie.s3.amazonaws.com/0dbaca258ebbe4fc6c6c4fb901b9cbc159d09c2b.png` and;
    1.2. The `\u2603` compared to the `☃` character.

Do we have the script or codes that generated the `asssessmentitems.json` from the .zip file mentioned above so we can change it and complete this PR?

**Comment by aronasorman** [comment]
Ah. So essentially we can't get the exact string from the mo file, because it differs slightly from what's in assessmentitems.json due to the url.

So the best solution in here is to do the same amazon.com to /content url conversion we do with the english assessment strings in the mo file, before we insert the mo file into the language pack.

That seems to be a central server issue though. Not sure if we have to change anything  here in this PR @cpauya.

File: kalite/topic_tools/__init__.py (hunk 2 above)
Comment (MCGallaspy) [inline kalite/topic_tools/__init__.py]: Hmm, you're right...

Comment (MCGallaspy) [inline kalite/topic_tools/__init__.py]:
```python
for d in item_data['hints']:
    d['content'] = _(d.get('content'))
```
This should work and is succint

**Comment by cpauya** [comment]
Yes @aronasorman - let's leave this open until https://github.com/fle-internal/ka-lite-central/issues/238 is done and merged then.

**Comment by jesumer** [comment]
Hi @aronasorman,

As what @cpauya hip chat instruction:
1. run `update_language_packs -l de --no-srts --no-dubbed` on central.
2. check an `*exercise-de.pot` contains the latest approved strings.
3. another option is, do an export of an `*exercise-de.pot` from `CrowdIn account of KA`.
4. compare that with the downloaded `*exercise-de.pot`.

Been tried to reproduce  the instruction above:

central1

but when I was at step 3 and trying to export I got this:
screen shot 2015-03-12 at 11 18 05 am

I guess my issue https://github.com/learningequality/ka-lite/issues/3087 is not necessary since we can now get the latest pot files from `crowdin`.

**Comment by aronasorman** [comment]
Let's merge this, so I can test it with the dummy language packs.

**Pull request merged**